fish-lsp: 1.0.10 -> 1.1.0

llakala opened the pull request:
We're hitting some changes upstream. I've gotten the package to actually build, but we're getting an error when running the binary. Any help would be appreciated!

Relevant error: 
```
node:internal/modules/cjs/loader:1210
  throw err;
  ^

Error: Cannot find module '@embedded_assets/fish_files/exec.fish'
Require stack:
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/virtual-fs.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/exec.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/process-env.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/env-manager.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/translation.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/document.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/parsing/symbol.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/parsing/set.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/parsing/barrel.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/node-types.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/tree-sitter.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/code-actions/quick-fixes.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/code-actions/code-action-handler.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/command.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/config.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/commander-cli-subcommands.js
- /nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/cli.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1207:15)
    at Module._load (node:internal/modules/cjs/loader:1038:27)
    at Module.require (node:internal/modules/cjs/loader:1289:19)
    at require (node:internal/modules/helpers:182:18)
    at Object.<anonymous> (/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/virtual-fs.js:45:27)
    at Module._compile (node:internal/modules/cjs/loader:1521:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1623:10)
    at Module.load (node:internal/modules/cjs/loader:1266:32)
    at Module._load (node:internal/modules/cjs/loader:1091:12)
    at Module.require (node:internal/modules/cjs/loader:1289:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/virtual-fs.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/exec.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/process-env.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/env-manager.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/translation.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/document.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/parsing/symbol.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/parsing/set.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/parsing/barrel.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/node-types.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/tree-sitter.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/code-actions/quick-fixes.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/code-actions/code-action-handler.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/command.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/config.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/utils/commander-cli-subcommands.js',
    '/nix/store/gld7i90szb9l2rizh3zmfq4fks4frmkm-fish-lsp-1.0.11/share/fish-lsp/out/cli.js'
  ]
}
```

I had to pin to nodejs_20 to get as far as I did, since node 22 and above give us this error:
```
SyntaxError: The requested module 'fs-extra' does not provide an export named 'copySync'
       >     at ModuleJob._instantiate (node:internal/modules/esm/module_job:228:21)
       >     at async ModuleJob.run (node:internal/modules/esm/module_job:337:5)
       >     at async onImport.tracePromise.__proto__ (node:internal/modules/esm/loader:651:26)
       >     at async asyncRunEntryPointWithESMLoader (node:internal/modules/run_main:117:5)
```

I also think the `postPatch` is now unnecessary - the `bin/fish-lsp` file seems to be removed from the repo. I don't want to remove it yet, though, until I can get the binary actually functioning.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

llakala commented:
cc @petertriho - nixpkgs won't send out review requests for draft PRs, but I'm just drafting till the package actually functions.

ndonfris commented:
Sorry, just now seeing this since I'm not a nix maintainer. 
 
The server should be working on [`v1.1.0`](https://github.com/ndonfris/fish-lsp/releases/tag/v1.1.0) and maybe the version tagged here? Unfortunately, it appears I was still getting the source maps stable on `v1.0.11` so tracking down those logs  might be tricky. 

___

To fix some of the vscode client extension's issues related to portability, I decided to bundle the whole server into a single commonjs file.

I tried to consider nix and some of the other package managers tracking the project when working on this solution. 

```sh
yarn dev # also related: `yarn build` 
```

Should build two different node dependent binaries of the server. 

- One with everything embedded (mostly for testing purposes on prereleases): `bin/fish-lsp`
- One that uses the modules installed from the `yarn install` command: `dist/fish-lsp` 
 
```
bin
├── fish-lsp.map
└── fish-lsp
dist
├── fish-lsp.d.ts
├── fish-lsp.map
└── fish-lsp 
```     

**NOTE:** symlinking to `dist/fish-lsp` is probably the ideal solution if possible. This allows error messages to use the source maps to point back to the origional `src/**/*.ts` locations causing issues.  

___
 
 > I also think the postPatch is now unnecessary - the bin/fish-lsp file seems to be removed from the repo. I don't want to remove it yet, though, until I can get the binary actually functioning.

Since the bundled build step is minifying and embedding wasm for tree-sitter into the outputted commonjs code, I removed those folders from git tracking. 

To me, the error mentioned here actually looks like it is caused by using [`out/cli.js`](https://github.com/NixOS/nixpkgs/pull/446470/files#diff-0751d1f3a4863265cab9db83269c2c8a4773cb13e47b3fafdccd86c5ec7610f1R60) instead of one of the bundled locations. The `out/` directory is now only used as an intermediate stage during the build process, to tree-shake the dependencies. 

I'm fairly confident that a similiar error came up for me when I was initially converting the test suite over to the new structure, which was resolved by mocking the embedded files with their actual content (esbuild handles this in `{dist,bin}/fish-lsp`). 

Another thing that might help is to comb through the `node_modules/` structure in [#447958](https://github.com/NixOS/nixpkgs/pull/447958). The vscode extension tagged in that PR uses  [`fish-lsp@1.1.0`](https://github.com/ndonfris/fish-lsp/releases/tag/v1.1.0) and appears to be working normally.

llakala committed:
fish-lsp: 1.0.10 -> 1.1.0

ndonfris commented on pkgs/by-name/fi/fish-lsp/package.nix:
```diff
@@ -39,32 +39,28 @@ stdenv.mkDerivation (finalAttrs: {
       --prefix PATH : "${
         lib.makeBinPath [
-          fish
+          fish # TODO: may be removable, seems to build and function fine without it
```
Oh that makes. Ya, you can probably remove it then.

ndonfris commented:
LGTM

jasonxue1 reviewed (changes requested):
In #454909 , @SuperSandro2000 said:
> We should rather use a condition based on nix code to notice breakages in the future

I think in this pr, it should be changed for same reason.

<img width="926" height="425" alt="image" src="https://github.com/user-attachments/assets/1eda656b-f1d3-4386-ac6d-786e08a03933" />

llakala commented:
For my PR, I just removed the file, since ndonfris said this:
> Since the bundled build step is minifying and embedding wasm for tree-sitter into the outputted commonjs code, I removed those folders from git tracking.

I don't think we need logic here for it.

llakala commented:
Sidenote @ndonfris: is there a way to stop the build time from being injected into the package for the `info` subcommand? As it stands, the package is currently non-reproducible because of this. Perhaps there's a nixy way to prevent this, but I don't know it.
